drv/stm32xx-sys/src/main.rs

```diff
@@ -255,20 +395,241 @@ impl idl::InOrderSysImpl for ServerImpl<'_> {
-        0
+        cfg_if! {
+            if #[cfg(feature = "exti")] {
+                notifications::EXTI_WILDCARD_IRQ_MASK
+            } else {
+                // We don't use notifications, don't listen for any.
+                0
+            }
+        }
     }
 
-    fn handle_notification(&mut self, _bits: u32) {
-        unreachable!()
+    fn handle_notification(&mut self, bits: u32) {
+        cfg_if! {
+            if #[cfg(feature = "exti")] {
+                if bits & notifications::EXTI_WILDCARD_IRQ_MASK != 0 {
+                    // Some combination of external pin change interrupts have
+                    // been triggered! Our first task is to determine which.
+                    // Fortunately, that's easy; the peripheral has a "pending"
+                    // register that mirrors the NVIC's, and latches events when
+                    // they occur.
+                    let pending = self.exti.cpupr1.read();
+                    // We'll cross-correlate that with the mask register so we
+                    // only deliver events a task is actually interested in.
+                    let enabled = self.exti.cpuimr1.read();
+
+                    // The EXTI bits corresponding to pin events are in the
+                    // lower 16 bits. The PAC didn't expect us to use these as a
+                    // group, so we must bypass it here manually:
+                    let pending = pending.bits() as u16;
+                    let enabled = enabled.bits() as u16;
+
+                    let pending_and_enabled = pending & enabled;
+
+                    let mut bits_to_acknowledge = 0u16;
+
+                    for pin_idx in 0..16 {
+                        if pending_and_enabled & 1 << pin_idx != 0 {
+                            // A channel is pending! We need to handle this
+                            // basically like the kernel handles native hardware
+                            // interrupts, which means
+                            // - Post the event to the owning task
+                            // - Mask the interrupt
+                            // - Clear the pending bit (we have to do this
+                            //   manually unlike native interrupts).
+
+                            if let Some((_, taskid, mask)) = generated::EXTI_DISPATCH_TABLE[pin_idx] {
+                                let taskid = sys_refresh_task_id(taskid);
+                                sys_post(taskid, mask);
+                            } else {
+                                // spurious interrupt.
+                                // TODO: probably add this to a counter; it's
+                                // almost certainly a bug, but panicking would
+                                // be rude since this is a shared service.
+                            }
+
+                            bits_to_acknowledge |= 1 << pin_idx;
+                        }
+                    }
```

Comment:
_very_ nitpicky, take it or leave it: i bet if we wrote
```suggestion
                    for (pin_idx, entry) in generated::EXTI_DISPATCH_TABLE.iter().enumerate() {
                        if pending_and_enabled & 1 << pin_idx != 0 {
                            // A channel is pending! We need to handle this
                            // basically like the kernel handles native hardware
                            // interrupts, which means
                            // - Post the event to the owning task
                            // - Mask the interrupt
                            // - Clear the pending bit (we have to do this
                            //   manually unlike native interrupts).

                            if let Some((_, taskid, mask)) = entry {
                                let taskid = sys_refresh_task_id(taskid);
                                sys_post(taskid, mask);
                            } else {
                                // spurious interrupt.
                                // TODO: probably add this to a counter; it's
                                // almost certainly a bug, but panicking would
                                // be rude since this is a shared service.
                            }

                            bits_to_acknowledge |= 1 << pin_idx;
                        }
                    }
```
here, the compiler wouldn't generate bounds checks every time we index the dispatch table...it might already be smart enough not to do that, though...

Reply:
I actually meant to change this to iter/enumerate before posting it, and I'm kind of surprised Clippy isn't flipping out about it. I guess the single use of `pin_idx` to construct a bitmask is enough to reassure it that it isn't _only_ being used to index an array.